File: python_system/python_shared/api/block_ad.py
"""
block_ad.py - Block Ad API Endpoint
Records when a user blocks an ad from appearing in their feed
Converted from PHP to Python
"""

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from pathlib import Path
from datetime import datetime
import json
import re

router = APIRouter()

# Storage paths
BLOCKED_ADS_PATH = Path(__file__).parent.parent / "data" / "blocked_ads"
BLOCKED_ADS_PATH.mkdir(parents=True, exist_ok=True)


@router.post("/block_ad")
async def block_ad(request: Request):
    """Block an ad from appearing in user's feed"""
    try:
        data = await request.json()

        ad_id = data.get("ad_id", "")
        if not ad_id:
            return JSONResponse(
                status_code=400,
                content={"success": False, "error": "Missing ad_id"}
            )

        reason = data.get("reason", "not_specified")
        comment = data.get("comment", "")[:500]  # Limit to 500 chars
        device_id = data.get("device_id", "unknown")
        timestamp = data.get("timestamp", int(datetime.now().timestamp() * 1000))

        # Sanitize inputs
        ad_id = re.sub(r'[^a-zA-Z0-9\-_]', '', ad_id)
        reason = re.sub(r'[^a-zA-Z0-9_]', '', reason)
        device_id = re.sub(r'[^a-zA-Z0-9\-_]', '', device_id)

        # Create block record
        block_record = {
            "ad_id": ad_id,
            "reason": reason,
            "comment": comment,
            "device_id": device_id,
            "ip_address": request.client.host if request.client else "unknown",
            "user_agent": request.headers.get("user-agent", "unknown"),
            "timestamp": timestamp,
            "created_at": datetime.now().isoformat()
        }

        # Save to file system (daily files)
        today = datetime.now().strftime("%Y-%m-%d")
        file_path = BLOCKED_ADS_PATH / f"{today}.json"

        daily_blocks = []
        if file_path.exists():
            with open(file_path, "r") as f:
                daily_blocks = json.load(f)

        daily_blocks.append(block_record)

        with open(file_path, "w") as f:
            json.dump(daily_blocks, f, indent=2)

        # Try to update database as well
        try:
            import sys
            sys.path.insert(0, str(Path(__file__).parent.parent))
            from database import SessionLocal
            from models import Ad

            db = SessionLocal()
            ad = db.query(Ad).filter(Ad.ad_id == ad_id).first()
            if ad:
                ad.blocks_count = (ad.blocks_count or 0) + 1
                db.commit()
            db.close()
        except Exception:
            pass  # Database update is optional

        return {
            "success": True,
            "message": "Ad blocked successfully",
            "ad_id": ad_id
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )


@router.get("/blocked_ads")
async def get_blocked_ads(date: str = None):
    """Get blocked ads for a specific date or all"""
    try:
        blocked_ads = []

        if date:
            # Get specific date
            file_path = BLOCKED_ADS_PATH / f"{date}.json"
            if file_path.exists():
                with open(file_path, "r") as f:
                    blocked_ads = json.load(f)
        else:
            # Get all blocked ads
            for file_path in BLOCKED_ADS_PATH.glob("*.json"):
                with open(file_path, "r") as f:
                    blocked_ads.extend(json.load(f))

        return {
            "success": True,
            "blocked_ads": blocked_ads,
            "total": len(blocked_ads)
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )

File: python_system/python_shared/api/block_ad.php
<?php
/**
 * Block Ad API Endpoint
 * Records when a user blocks an ad from appearing in their feed
 */

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    echo json_encode(['success' => false, 'error' => 'Method not allowed']);
    exit;
}

// Get JSON input
$input = json_decode(file_get_contents('php://input'), true);

if (!$input || empty($input['ad_id'])) {
    http_response_code(400);
    echo json_encode(['success' => false, 'error' => 'Missing ad_id']);
    exit;
}

$adId = $input['ad_id'];
$reason = $input['reason'] ?? 'not_specified';
$comment = $input['comment'] ?? '';
$deviceId = $input['device_id'] ?? 'unknown';
$timestamp = $input['timestamp'] ?? time() * 1000;

// Sanitize inputs
$adId = preg_replace('/[^a-zA-Z0-9\-_]/', '', $adId);
$reason = preg_replace('/[^a-zA-Z0-9_]/', '', $reason);
$comment = substr(strip_tags($comment), 0, 500);
$deviceId = preg_replace('/[^a-zA-Z0-9\-_]/', '', $deviceId);

// Define storage paths
$basePath = dirname(__DIR__) . '/data/blocked_ads';
$dbPath = dirname(__DIR__) . '/shared/database/adsphere.db';

// Create directory if not exists
if (!is_dir($basePath)) {
    mkdir($basePath, 0755, true);
}

// Create block record
$blockRecord = [
    'ad_id' => $adId,
    'reason' => $reason,
    'comment' => $comment,
    'device_id' => $deviceId,
    'ip_address' => $_SERVER['REMOTE_ADDR'] ?? 'unknown',
    'user_agent' => $_SERVER['HTTP_USER_AGENT'] ?? 'unknown',
    'timestamp' => $timestamp,
    'created_at' => date('c')
];

// Save to file system (for quick access)
$filePath = $basePath . '/' . date('Y-m-d') . '.json';
$dailyBlocks = [];

if (file_exists($filePath)) {
    $dailyBlocks = json_decode(file_get_contents($filePath), true) ?? [];
}

$dailyBlocks[] = $blockRecord;
file_put_contents($filePath, json_encode($dailyBlocks, JSON_PRETTY_PRINT), LOCK_EX);

// Also save to SQLite database if available
try {
    if (file_exists($dbPath)) {
        $db = new PDO('sqlite:' . $dbPath);
        $db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

        // Create blocked_ads table if not exists
        $db->exec("CREATE TABLE IF NOT EXISTS blocked_ads (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ad_id TEXT NOT NULL,
            reason TEXT,
            comment TEXT,
            device_id TEXT,
            ip_address TEXT,
            user_agent TEXT,
            created_at DATETIME DEFAULT CURRENT_TIMESTAMP
        )");

        // Insert record
        $stmt = $db->prepare("INSERT INTO blocked_ads (ad_id, reason, comment, device_id, ip_address, user_agent) VALUES (?, ?, ?, ?, ?, ?)");
        $stmt->execute([
            $adId,
            $reason,
            $comment,
            $deviceId,
            $_SERVER['REMOTE_ADDR'] ?? '',
            $_SERVER['HTTP_USER_AGENT'] ?? ''
        ]);

        // Update ad block count
        $db->exec("UPDATE ads SET blocks_count = COALESCE(blocks_count, 0) + 1 WHERE ad_id = " . $db->quote($adId));
    }
} catch (PDOException $e) {
    error_log('Block ad database error: ' . $e->getMessage());
}

// Return success
echo json_encode([
    'success' => true,
    'message' => 'Ad blocked successfully',
    'ad_id' => $adId
]);

